arbnode/seq_coordinator.go

```diff
@@ -671,46 +675,68 @@ func (c *SeqCoordinator) Start(ctxIn context.Context) {
 	}
 }
 
-func (c *SeqCoordinator) waitForHandoff(ctx context.Context) string {
-	var nextChosen string
+// Calls check() on retryInterval until it returns a non-default value, which is then returned.
+// On context timeout, this returns the default value.
+func waitFor[T comparable](ctx context.Context, retryInterval time.Duration, check func() T) T {
+	var empty T
 	for {
-		var err error
-		nextChosen, err = c.CurrentChosenSequencer(ctx)
-		if err == nil && nextChosen != "" && nextChosen != c.config.MyUrl() {
-			return nextChosen
+		result := check()
+		if result != empty {
+			return result
 		}
 		select {
 		case <-ctx.Done():
-			return ""
-		case <-time.After(c.config.RetryInterval):
+			return empty
+		case <-time.After(retryInterval):
 		}
 	}
 }
 
 func (c *SeqCoordinator) PrepareForShutdown() {
+	// normal context will be closed, use parent context
+	parentCtx := c.StopWaiter.GetParentContext()
+	handoffCtx, cancel := context.WithTimeout(c.StopWaiter.GetContext(), c.config.ShutdownHandoffTimeout)
+	defer cancel()
+	if c.config.ShutdownHandoffTimeout != time.Duration(0) {
```

Comment:
I'm thinking this should only be done if c.CurrentlyChosen()
This opens us to a small race condition which shouldn't really matter (if we get chosen during StopAndWait correctness isn't compromised).
Another option: only print the log waiting after checking once that there isn't another live sequencer.

Reply:
Good catch. I think the race condition is fine because of how unlikely it is. In theory we could kill our own liveliness earlier, but keep extending our lockout, but I think it's better to keep the code simpler unless we actually hit this race condition.